feat(testRunner): UnifiedTestRunner and friends

VijayKrishna opened the pull request:
1. Setting up the UnifiedTestRunner that will rely on the JUnit
   Platform to autodiscover the execute tests across Junit 5/4/3 and
   TestNG.
   a. Issuing separate code paths for test discovery (`containsExecutableTest`)
       and test execution (`getExecutableTest`).
   b. Using a different discovery path is useful to answer questions questions like:
       "Does the testClass have executable tests?", and as such,
       "Can we use the UnifiedTestRunner?"
   c. If the discovery path fails, then the UnifiedTest runner is not viable for
       a given test class.
2. We give the UnifiedTestRunner first dibs as part of the TestType discovery
     process in AbstractTestRunner.
3. UnifiedTestListenerAdapter is a TestExecutionListener (Junit Platform)
     implementation  that hosts a ITacocoTestLister implementation.
     This is the current approach to listening test events from a variety
     of test listening implementations (Junit X, TestNG, etc).
4. Updating pom.xml to pull in a compile time dependency on JUnit Platform
    a. Also adding runtime dependencies on test engines that confrom to the JUnit
        Platform spec.

Next Steps: integration tests that run on projects.

VijayKrishna committed:
feat(testRunner): UnifiedTestRunner and friends

1. Setting up the UnifiedTestRunner that will rely on the JUnit
   Platform to autodiscover the execute tests across Junit 5/4/3 and
   TestNG.
   a. Issuing separate code paths for test discovery (`containsExecutableTest`)
       and test execution (`getExecutableTest`).
   b. Using a different discovery path is useful to answer questions questions like:
       "Does the testClass have executable tests?", and as such,
       "Can we use the UnifiedTestRunner?"
   c. If the discovery path fails, then the UnifiedTest runner is not viable for
       a given test class.
2. We give the UnifiedTestRunner first dibs as part of the TestType discovery
     process in AbstractTestRunner.
3. UnifiedTestListenerAdapter is a TestExecutionListener (Junit Platform)
     implementation  that hosts a ITacocoTestLister implementation.
     This is the current approach to listening test events from a variety
     of test listening implementations (Junit X, TestNG, etc).
4. Updating pom.xml to pull in a compile time dependency on JUnit Platform
    a. Also adding runtime dependencies on test engines that confrom to the JUnit
        Platform spec.

Next Steps: integration tests that run on projects.

VijayKrishna commented:
@kajdreef i have not setup the tests yet. Will get the ball rolling on that shortly /within this very PR/. In the meanwhile, feel free to try this out locally and see if this works for your the projects that you were trying Tacoco with.

kajdreef commented:
When running the changes on the TestNG project it fails, see output here [output_spiderMath_TestNG.txt](https://github.com/spideruci/tacoco/files/4441304/output_spiderMath_TestNG.txt). maybe for now we should consider TestNG first (not using testng-junit5), and then let it check for JUnit4 and JUnit5 through the unified approach. 

Interestingly, if I update the TestNG dependency for spiderMath_TestNG to 7.0.0 the errors go away and I can run the project without a problem.

I will continue testing some things out today.

kajdreef commented on pom.xml:
```diff
@@ -147,9 +149,37 @@
+			<version>${junit.jupiter.version}</version>
+			<scope>runtime</scope>
+		</dependency>
+		<dependency>
+			<groupId>com.github.testng-team</groupId>
+			<artifactId>testng-junit5</artifactId>
+			<version>0.0.1</version>
+			<scope>runtime</scope>
+		</dependency>
```
I was looking at this again, and I am not sure if we should support TestNG at the moment through the unified platform. It seems that this tool is not officially supported by TestNG, so I am curious how this will continue to be developed in the future.

VijayKrishna replied:
If we just remove this runtime dependency, we should be able to by-pass the Unified Engine? The other alternative is to wrap all this crap in try-blocks and try each applicable test type (unified, junit3, junit4, testng) and take the one that works correctly the first time around.

VijayKrishna replied:
fabd838ccc5edffb72731718311fd111656eb70f adds necessary exception handling early in the test-type-recognition phase to be able to avoid the kind of exceptions that you caught. Additionally, i took the trouble of adding other try-catch blocks so that when running Tacoco as part of an experiment, it does not take the whole experiment's Python/Bash script down with it.

VijayKrishna replied:
the try-catch  in containsExecutableTest (as part of fabd838ccc5edffb72731718311fd111656eb70f) will ensure that it returns `false` if there is any issue in correctly discovering a test in a given testClass. I might even add a "dry run" mechanism to try running a discovered test from the testClass to see if the discovery actually discovered an executable test, or if it just discovered something that seemed like a test but really it was nothing.

VijayKrishna replied:
I went ahead and setup the basic bits for an optional test dry run in 3b59e09bca6ac6d0df6e651857acefad3f4bbd4f. We can do the necessary plumbing to make sure that this is configurable via command line options or Tacoco's config file.

VijayKrishna committed:
Handling unforeseen exception in UnifiedTestRunner

VijayKrishna committed:
Optional test dryRuns:
Setting up the bits necessary to start doing dryruns when figuring out if a
testClass has actual *executable* tests.

VijayKrishna committed:
Integration testing w/ spideruci/projects4testing
- bringing back integration-test in pom.xml
- resurrecting and making it  aware of projects4testing
- updating pom.xml to: pull spideruci/projects4testing; run tacoco-diag.

VijayKrishna committed:
Path Update: spiderMath_JUnit -> spiderMath_JUnit4

VijayKrishna committed:
No need to copy projects4testing:
- projects4testing is cloned in tacoco_home.

VijayKrishna committed:
Cloning spideruci/projects4testing before_install

kajdreef commented on pom.xml:
```diff
@@ -200,6 +230,22 @@
 					</execution>
 				</executions>
 			</plugin>
+			<plugin>
+				<groupId>org.codehaus.mojo</groupId>
+				<artifactId>exec-maven-plugin</artifactId>
+				<version>1.4.0</version>
+				<executions>
+					<execution>
+						<phase>integration-test</phase>
+						<goals>
+							<goal>exec</goal>
+						</goals>
+						<configuration>
+							<executable>${basedir}/tacoco-diagnoses</executable>
+						</configuration>
+					</execution>
+				</executions>
+			</plugin>
```
Looks good! At the moment we have to clone the repository by hand, maybe we should include that in the tacoco_diagnoses?

VijayKrishna replied:
I want the integration phase to be TravisCI-first. So i am comfortable with this lopsided approach for now -- unless you think that running `mvn integration-test` is key? Was it part of `mvn install` or `mvn publish`?

VijayKrishna replied:
Github now hosts Maven packages. And it enables you to consume Maven packages much like Maven Central. If anything, i am guessing we can setup projects4testing to be a non-Java (or vanila) Maven package that does not compile, or run or anything... and simply setup us such a package as a `integration` phase dependency for Tacoco.

kajdreef replied:
If you run mvn install locally it will also run the integration-test phase, so locally will fail if you didn't clone the projects4testing yet, so how do we want to tackle that?

VijayKrishna replied:
see my other comment. Passing simple -DskipXYZ flags to mvn install does the trick.

kajdreef commented on tacoco-diagnoses:
```diff
@@ -0,0 +1,37 @@
+tacoco_home=$(pwd)
+
+ls -alh .
+ls -alh $tacoco_home/projects4testing/
```
Should we add the git clone command here? 

```suggestion
ls -alh .
git clone https://github.com/spideruci/projects4testing
ls -alh $tacoco_home/projects4testing/
```

VijayKrishna replied:
I tried adding the git clone setup in the travis script, and it did not perform the clone. Which is why i moved the `git clone` to the `before_install` section of `.travis.yml` See 92411e44f6e0eb081c8748bdf20e97075d943bb7

kajdreef commented on tacoco-diagnoses:
```diff
@@ -0,0 +1,37 @@
+    echo "Tacoco failed with spiderMath_JUnit4..."
+    exit 1
+fi
+
```
Do we also want to remove the projects4testing in the end? or just add it to the .gitignore so now one will by accident commit anything or add it as a submodule.

VijayKrishna replied:
I thought this was pretty much self-contained in a TravisCI instance? so we should not have to worry about it getting added auto-magically to the repo? If we wanted to add this as a submodule (git) that would be a different conversation.

kajdreef replied:
I should have clarified. For travis this is okay, however, integration-test is also run locally when someone runs mvn install or mvn package. Without projects4testing this phase will fail.

VijayKrishna replied:
`mvn install -DskipTests -DskiptITs` // (assuming that i got the args right) this will skip both unit tests and the integration-test phase. Making this the default, and configuring it specifically for TravisCI should be a different change. Passing those flags on the command line in the meanwhile is not a terrible workaround.

VijayKrishna committed:
Testing the launcher

VijayKrishna committed:
spelling: analyzer

VijayKrishna committed:
Adding spiderMath_JUnit4_single_module

VijayKrishna committed:
Disabling mvn -Panalyzer for now.
